docs/integrations/custom_middleware.md

```diff
@@ -0,0 +1,191 @@
+# Writing your own Custom Middleware
+
+## How to write your own custom middleware for Mirascope
+
+`middleware_decorator` is a helper function to assist in helping you wrap any Mirascope call.
+We will be creating an example decorator `with_saving` that saves some metadata after a Mirascope call:
+
+### Writing the decorator
+
+```python
+from mirascope.integrations import middleware_decorator
+
+def with_saving(fn):
+    """Saves some data after a Mirascope call."""
+
+    return middleware_decorator(
+        fn,
+        custom_context_manager=custom_context_manager,
+        custom_decorator=custom_decorator,
+        handle_call_response=handle_call_response,
+        handle_call_response_async=handle_call_response_async,
+        handle_stream=handle_stream,
+        handle_stream_async=handle_stream_async,
+        handle_response_model=handle_response_model,
+        handle_response_model_async=handle_response_model_async,
+        handle_structured_stream=handle_structured_stream,
+        handle_structured_stream_async=handle_structured_stream_async,
+    )
+```
+
+We will go over each of the different functions.
+
+### `handle_call_response` and `handle_call_response_async`
+
+These functions will be called after making a mirascope call with the following signature:
+
+```python
+from typing import Callable
+
+from mirascope.core.base import BaseCallResponse
+
+
+def handle_call_response(
+    result: BaseCallResponse, fn: Callable, context_manager: ContextManagerType | None
+):
+    # handle after call here
```

Comment:
I think we should show an implementation for at least one of the functions (if not all) even if it's something super simple.

Reply:
See my example with SQLModel, you think I should move that code snippet up here? Should I do the same with the other handlers, or extrapolate.

Reply:
same with other handlers, and yeah I think if we're going to demonstrate how to build your own by implementing `with_saving` we should essentially implement with saving and do the same for other handlers